README.md
<h1 align="center">Home Dashboard Train Manager </h1>
<p align="center">
    <img src="https://travis-ci.org/iamtomhewitt/home-dashboard-train-manager.svg"/>
    <img src="https://heroku-badge.herokuapp.com/?app=home-dashboard-train-manager&style=round&svg=1"/>
</p>
<p align="center">
    A Node Express app for piggy backing the api of <a href="https://github.com/jpsingleton/Huxley">Huxley</a> to get train timetables for my <a href="https://github.com/iamtomhewitt/home-dashboard">home dashboard</a>. This project basically just reformats the JSON response from Huxley into a more basic model for my dashboard.
</p>
<p align="center">🙌 Big thanks to <a href="https://github.com/jpsingleton">James Singleton</a> for making Huxley!</p>

## Pipeline
* `Travis` tests the repo using `npm test`, which runs `mocha 'tests/**/*.js' --exit`
* Once the tests pass, `Heroku` deploys the app.
* When the app is deployed, you can make requests to it [here](https://home-dashboard-train-manager.herokuapp.com/).
* You will need an API key to make requests. You can get one [here](http://realtime.nationalrail.co.uk/OpenLDBWSRegistration/Registration).

## Endpoints

### `/ (GET)`
The root endpoint, returning information about the app.

#### Responses
* `200` success
```json
{
    "status": "🚂 SERVER OK",
    "version": "1.0.0",
    "endpoints": [
        {
            "path": "/",
            "methods": [
                "GET"
            ]
        },
        {
            "path": "/departures",
            "methods": [
                "GET"
            ]
        },
        {
            "path": "/arrivals",
            "methods": [
                "GET"
            ]
        }
    ]
}
```

### `/departures (GET)`
* Returns all the departures for a given station. You can find the station codes [here](https://www.nationalrail.co.uk/stations_destinations/48541.aspx).
* Query parameters:
	* `stationCode=<code>`
	* `numberOfResults=<number>`
	* `apiKey=<your api key>`

#### Responses
* `200` success
```json
{
    "status": 200,
    "station": "Leeds",
    "stationCode": "LDS",
    "timetable": [
        {
            "destination": "Doncaster",
            "scheduledDepartTime": "2020-01-14T12:21:00.000Z",
            "actualDepartTime": "2020-01-14T12:21:00.000Z",
            "platform": "2C",
            "cancelled": false,
            "busReplacement": false
        },
        {
            "destination": "Bradford Forster Square",
            "scheduledDepartTime": "2020-01-14T12:42:00.000Z",
            "actualDepartTime": "2020-01-14T12:42:00.000Z",
            "platform": "3B",
            "cancelled": false,
            "busReplacement": false
        },
        {
            "destination": "London Kings Cross",
            "scheduledDepartTime": "2020-01-14T12:45:00.000Z",
            "actualDepartTime": "2020-01-14T12:45:00.000Z",
            "platform": "9",
            "cancelled": false,
            "busReplacement": false
        },
        {
            "destination": "Manchester Airport",
            "scheduledDepartTime": "2020-01-14T12:45:00.000Z",
            "actualDepartTime": "2020-01-14T12:45:00.000Z",
            "platform": "16A",
            "cancelled": false,
            "busReplacement": false
        },
        {
            "destination": "Leeds Bradford Airport (Bus)",
            "scheduledDepartTime": "2020-01-14T13:09:00.000Z",
            "actualDepartTime": "2020-01-14T13:09:00.000Z",
            "platform": null,
            "cancelled": false,
            "busReplacement": true
        }
    ]
}
```

* `400` if there was a problem with query parameters 
```json
{
    "status": 400,
    "message": "There are missing query parameters"
}
```

* `500` error
```json
{
    "status": 500,
    "message": "<error message>"
}
```

### `/arrivals (GET)`
* Returns all the arrivals for a given station. You can find the station codes [here](https://www.nationalrail.co.uk/stations_destinations/48541.aspx).
* Query parameters:
	* `stationCode=<code>`
	* `numberOfResults=<number>`
	* `apiKey=<your api key>`

#### Responses
* `200` success
```json
{
    "status": 200,
    "station": "Leeds",
    "stationCode": "LDS",
    "timetable": [
        {
            "origin": "London Kings Cross",
            "scheduledArrivalTime": "2020-01-14T12:48:00.000Z",
            "actualArrivalTime": "2020-01-14T12:48:00.000Z",
            "platform": "6",
            "cancelled": false,
            "busReplacement": false
        },
        {
            "origin": "Liverpool Lime Street",
            "scheduledArrivalTime": "2020-01-14T12:49:00.000Z",
            "actualArrivalTime": "2020-01-14T12:49:00.000Z",
            "platform": "11D",
            "cancelled": false,
            "busReplacement": false
        },
        {
            "origin": "Halifax",
            "scheduledArrivalTime": "2020-01-14T12:52:00.000Z",
            "actualArrivalTime": "2020-01-14T12:52:00.000Z",
            "platform": "8B",
            "cancelled": false,
            "busReplacement": false
        },
        {
            "origin": "Knaresborough",
            "scheduledArrivalTime": "2020-01-14T12:53:00.000Z",
            "actualArrivalTime": "2020-01-14T12:53:00.000Z",
            "platform": null,
            "cancelled": false,
            "busReplacement": false
        },
        {
            "destination": "Leeds",
            "scheduledArrivalTime": "2020-01-14T13:05:00.000Z",
            "actualArrivalTime": "2020-01-14T13:05:00.000Z",
            "platform": null,
            "cancelled": false,
            "busReplacement": true
        }
    ]
}
```

* `400` if there was a problem with query parameters 
```json
{
    "status": 400,
    "message": "There are missing query parameters"
}
```

* `500` error
```json
{
    "status": 500,
    "message": "<error message>"
}
```

CHANGELOG.md
## 1.0.6 (06/09/2020) 


### Issues in this release:

* [#3](https://github.com/iamtomhewitt/home-dashboard-train-manager/issues/3) - Use github actions



## 1.0.5 (14/07/2020) 


There are no issues in this release.

tests/tests.js
const request = require('supertest');
const assert = require('assert');
require('dotenv').config();

describe('/ tests', () => {
    let server;
    const apiKey = process.env.API_KEY;

    before(() => {
        server = require('../app').listen(3002);
    });

    after(() => {
        server.close();
    });

    it('/ gives 200', (done) => {
        request(server)
            .get('/')
            .send({
                stationCode: 'LDS',
                numberOfResults: 10,
                apiKey,
            })
            .expect(200)
            .end((err, response) => {
                if (err) {
                    return done(err);
                }
                const { version } = require('../package.json');
                const status = '🚂 SERVER OK';

                assert.equal(response.body.version, version);
                assert.equal(response.body.status, status);

                return done();
            });
    });

    it('/departures gives 200', (done) => {
        request(server)
            .get(`/departures?stationCode=LDS&numberOfResults=10&apiKey=${apiKey}`)
            .expect(200)
            .end((err, response) => {
                if (err) {
                    assert.fail(err.message);
                    return done(err);
                }

                assert.notEqual(response.body.timetable, null);
                assert.equal(response.body.timetable.length, 10);
                assert.equal(response.body.stationCode, 'LDS');

                return done();
            });
    });

    it('/departures gives 400 when parameters are missing', (done) => {
        request(server)
            .get('/departures')
            .expect(400, done);
    });

    it('/departures gives 500 when using an incorrect station', (done) => {
        request(server)
            .get(`/departures?stationCode=XYZ&numberOfResults=10&apiKey=${apiKey}`)
            .expect(500, done);
    });

    it('/arrivals gives 200', (done) => {
        request(server)
            .get(`/arrivals?stationCode=LDS&numberOfResults=10&apiKey=${apiKey}`)
            .expect(200)
            .end((err, response) => {
                if (err) {
                    return done(err);
                }

                assert.notEqual(response.body.timetable, null);
                assert.equal(response.body.timetable.length, 10);
                assert.equal(response.body.stationCode, 'LDS');

                return done();
            });
    });

    it('/arrivals gives 400 when parameters are missing', (done) => {
        request(server)
            .get('/arrivals')
            .expect(400, done);
    });

    it('/arrivals gives 500 when using an incorrect station', (done) => {
        request(server)
            .get(`/arrivals?stationCode=XYZ&numberOfResults=10&apiKey=${apiKey}`)
            .expect(500, done);
    });
});


app.js
const express = require('express');
const bodyParser = require('body-parser');
const listEndpoints = require('express-list-endpoints');
const request = require('request');
const { version } = require('./package.json');

const app = express();

const success = 200;
const badRequest = 400;
const serverError = 500;

app.use(bodyParser.json());
app.use(bodyParser.urlencoded({ extended: true }));

app.get('/', (req, res) => {
    res.status(success).send({ status: '🚂 SERVER OK', version, endpoints: listEndpoints(app) });
});

app.get('/departures', (req, res) => {
    const { stationCode } = req.query;
    const { apiKey } = req.query;
    const { numberOfResults } = req.query;

    let response;

    if (!stationCode || !apiKey || !numberOfResults) {
        response = {
            status: badRequest,
            message: 'There are missing query parameters',
        };
        res.status(badRequest).send(response);
        return;
    }

    const url = `https://huxley.apphb.com/departures/${stationCode}/${numberOfResults}?accessToken=${apiKey}`;

    request(url, (err, resp) => {
        if (err) {
            response = {
                status: serverError,
                message: err.message,
            };
            res.status(serverError).send(response);
            return;
        }

        if (resp.statusCode === 500) {
            response = {
                status: serverError,
                message: `There was an error from Huxley, perhaps the station code is incorrect (you sent '${stationCode}')`,
            };
            res.status(serverError).send(response);
            return;
        }

        const today = new Date(Date.now()).toLocaleString().slice(0, 10);
        const huxleyResponse = JSON.parse(resp.body);
        const stationName = huxleyResponse.locationName;
        const timetable = [];

        if (huxleyResponse.trainServices !== null) {
            huxleyResponse.trainServices.forEach((train) => {
                const scheduledDepartTime = new Date(`${today} ${train.std}`);
                const actualDepartTime = train.etd === 'On time' ? new Date(`${today} ${train.std}`) : new Date(`${today} ${train.etd}`);

                timetable.push({
                    destination: train.destination[0].locationName,
                    scheduledDepartTime,
                    actualDepartTime,
                    platform: train.platform,
                    cancelled: train.isCancelled,
                    busReplacement: false,
                });
            });
        }

        if (huxleyResponse.busServices !== null) {
            huxleyResponse.busServices.forEach((bus) => {
                const scheduledDepartTime = new Date(`${today} ${bus.std}`);
                const actualDepartTime = scheduledDepartTime;

                timetable.push({
                    destination: bus.destination[0].locationName,
                    scheduledDepartTime,
                    actualDepartTime,
                    platform: bus.platform,
                    cancelled: false,
                    busReplacement: true,
                });
            });
        }

        // Sort the timetable based on departure time so buses are not lumped at the end
        timetable.sort((a, b) => a.scheduledDepartTime.toString().localeCompare(b.scheduledDepartTime.toString()));

        response = {
            status: success,
            station: stationName,
            stationCode,
            timetable,
        };

        res.status(success).send(response);
    });
});

app.get('/arrivals', (req, res) => {
    const { stationCode } = req.query;
    const { apiKey } = req.query;
    const { numberOfResults } = req.query;

    let response;

    if (!stationCode || !apiKey || !numberOfResults) {
        response = {
            status: badRequest,
            message: 'There are missing query parameters',
        };
        res.status(badRequest).send(response);
        return;
    }

    const url = `https://huxley.apphb.com/arrivals/${stationCode}/${numberOfResults}?accessToken=${apiKey}`;

    request(url, (err, resp) => {
        if (err) {
            response = {
                status: serverError,
                message: err.message,
            };
            res.status(serverError).send(response);
            return;
        }

        if (resp.statusCode === 500) {
            response = {
                status: serverError,
                message: `There was an error from Huxley, perhaps the station code is incorrect (you sent '${stationCode}')`,
            };
            res.status(serverError).send(response);
            return;
        }

        const today = new Date(Date.now()).toLocaleString().slice(0, 10);
        const huxleyResponse = JSON.parse(resp.body);
        const stationName = huxleyResponse.locationName;
        const timetable = [];

        if (huxleyResponse.trainServices !== null) {
            huxleyResponse.trainServices.forEach((train) => {
                const scheduledArrivalTime = new Date(`${today} ${train.sta}`);
                const actualArrivalTime = train.eta === 'On time' ? new Date(`${today} ${train.sta}`) : new Date(`${today} ${train.eta}`);

                timetable.push({
                    origin: train.origin[0].locationName,
                    scheduledArrivalTime,
                    actualArrivalTime,
                    platform: train.platform,
                    cancelled: train.isCancelled,
                    busReplacement: false,
                });
            });
        }

        if (huxleyResponse.busServices !== null) {
            huxleyResponse.busServices.forEach((bus) => {
                const scheduledArrivalTime = new Date(`${today} ${bus.sta}`);
                const actualArrivalTime = scheduledArrivalTime;

                timetable.push({
                    destination: bus.destination[0].locationName,
                    scheduledArrivalTime,
                    actualArrivalTime,
                    platform: bus.platform,
                    cancelled: false,
                    busReplacement: true,
                });
            });
        }

        // Sort the timetable based on arrival time so buses are not lumped at the end
        timetable.sort((a, b) => a.scheduledArrivalTime.toString().localeCompare(b.scheduledArrivalTime.toString()));

        response = {
            status: success,
            station: stationName,
            stationCode,
            timetable,
        };

        res.status(success).send(response);
    });
});

const port = 3001;
app.listen(process.env.PORT || port, () => { });

module.exports = app;
